flink-metrics/flink-metrics-statsd/src/main/java/org/apache/flink/metrics/statsd/StatsDReporter.java

```diff
@@ -179,41 +254,130 @@ private String prefix(String ... names) {
+					insertFiller = true;
+				}
+			}
+		}
 
-				default:
-					if (chars != null) {
-						chars[pos] = c;
-					}
-					pos++;
+		if (chars == null) {
+			if (strLen > limit) {
+				return input.substring(0, limit);
+			} else {
+				return input; // happy path, input is entirely valid and under the limit
 			}
+		} else {
+			return new String(chars, 0, pos);
 		}
+	}
 
-		return chars == null ? input : new String(chars, 0, pos);
+	/**
+	 * filterCharacters() is called on each delimited segment of the metric.
+	 *
+	 * <p>We might get a string that has coded structures, references to instances of serializers and reducers, and even if
+	 * we normalize all the odd characters could be overly long for a metric name, likely to be truncated downstream.
+	 * Our choices here appear to be either to discard invalid metrics, or to pragmatically handle each of the various
+	 * issues and produce something that might be useful in aggregate even though the named parts are hard to read.
+	 *
+	 * <p>This function will find and remove all object references like @abcd0123, so that task and operator names are stable.
+	 * The name of an operator should be the same every time it is run, so we should ignore object hash ids like these.
+	 *
+	 * <p>If the segment is a tm_id, task_id, job_id, task_attempt_id, we can optionally trim those to the first 8 chars.
+	 * This can reduce overall length substantially while still preserving enough to distinguish metrics from each other.
+	 *
+	 * <p>If the segment is 50 chars or longer, we will compress it to avoid truncation. The compression will look like the
+	 * first 10 valid chars followed by a hash of the original.  This sacrifices readability for utility as a metric, so
+	 * that latency metrics might survive with valid and useful dimensions for aggregation, even if it is very hard to
+	 * reverse engineer the particular operator name.  Application developers can of course still supply their own names
+	 * and are not forced to rely on the defaults.
+	 *
+	 * <p>This will turn something like:
+	 * 		"TriggerWindow(TumblingProcessingTimeWindows(5000), ReducingStateDescriptor{serializer=org.apache.flink.api.java
+	 * 		.typeutils.runtime.PojoSerializer@f3395ffa, reduceFunction=org.apache.flink.streaming.examples.socket.
+	 * 		SocketWindowWordCount$1@4201c465}, ProcessingTimeTrigger(), WindowedStream.reduce(WindowedStream.java-301))"
+	 *
+	 * <p>into:  "TriggerWin_c2910b88"
+	 */
+	@Override
+	public String filterCharacters(String input) {
+		// remove instance references
+		Matcher hasRefs = instanceRef.matcher(input);
+		if (hasRefs.find()) {
+			input = hasRefs.replaceAll("");
+		}
+		// compress segments too long
+		if (input.length() >= 50) {
+			return filterNCharacters(input, 10) + "_" + Integer.toHexString(input.hashCode());
+		}
+		int limit = Integer.MAX_VALUE;
+		// optionally shrink flink ids
+		if (shortIds && input.length() == 32 && flinkId.matcher(input).matches()) {
```

Comment:
Why truncate the IDs? This could could create collisions and lead to invalid metrics. It also makes it impossible to take the ID out of a metric and query the API for more information about that job. This makes it very hard to work with.